packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -123,6 +123,56 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
     properties.add(new DiagnosticsProperty<Object>('owner', owner, level: parent != null ? DiagnosticLevel.hidden : DiagnosticLevel.info, defaultValue: null));
     properties.add(new DiagnosticsProperty<dynamic>('creator', debugCreator, defaultValue: null, level: DiagnosticLevel.debug));
   }
+
+  static void _clipAndPaint<T>(Canvas canvas, Clip clipBehavior, T t, Rect bounds, void painter()) {
+    void Function(bool doAA) canvasClipCall;
+    if (t is Path) {
+      canvasClipCall = (bool doAA) => canvas.clipPath(t, doAntiAlias: doAA);
+    } else if (t is RRect) {
+      canvasClipCall = (bool doAA) => canvas.clipRRect(t, doAntiAlias: doAA);
+    } else if (t is Rect) {
+      canvasClipCall = (bool doAA) => canvas.clipRect(t, doAntiAlias: doAA);
+    }
+    assert(canvasClipCall != null); // t should be one of Path, RRect, Rect
+    canvas.save();
+    switch (clipBehavior) {
+      case Clip.none:
+        break;
+      case Clip.hardEdge:
+        canvasClipCall(false);
+        break;
+      case Clip.antiAlias:
+        canvasClipCall(true);
+        break;
+      case Clip.antiAliasWithSaveLayer:
+        canvasClipCall(true);
+        canvas.saveLayer(bounds, new Paint());
+        break;
+    }
+    painter();
+    if (clipBehavior == Clip.antiAliasWithSaveLayer) {
+      canvas.restore();
+    }
+    canvas.restore();
+  }
+
+  /// Clip [Canvas] with [Path] according to [Clip] and then paint. [Canvas] is
+  /// restored to the pre-clip status afterwards.
+  static void clipPathAndPaint(Canvas canvas, Clip clipBehavior, Path path, Rect bounds, void painter()) {
+    _clipAndPaint(canvas, clipBehavior, path, bounds, painter);
+  }
+
+  /// Clip [Canvas] with [Path] according to [RRect] and then paint. [Canvas] is
+  /// restored to the pre-clip status afterwards.
+  static void clipRRectAndPaint(Canvas canvas, Clip clipBehavior, RRect rrect, Rect bounds, void painter()) {
+    _clipAndPaint(canvas, clipBehavior, rrect, bounds, painter);
+  }
+
+  /// Clip [Canvas] with [Path] according to [Rect] and then paint. [Canvas] is
+  /// restored to the pre-clip status afterwards.
+  static void clipRectAndPaint(Canvas canvas, Clip clipBehavior, Rect rect, Rect bounds, void painter()) {
+    _clipAndPaint(canvas, clipBehavior, rect, bounds, painter);
+  }
```

Comment:
why are these on `Layer`? Seems like an odd place for them. Since they don't do anything with layers, i'd put them in the `painting` layer of our system.